synapse/storage/databases/main/relations.py

```diff
@@ -13,26 +13,63 @@
 # limitations under the License.
 
 import logging
-from typing import List, Optional, Tuple, Union, cast
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Collection,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
 
 import attr
+from frozendict import frozendict
 
 from synapse.api.constants import RelationTypes
 from synapse.events import EventBase
 from synapse.storage._base import SQLBaseStore
-from synapse.storage.database import LoggingTransaction, make_in_list_sql_clause
+from synapse.storage.database import (
+    DatabasePool,
+    LoggingDatabaseConnection,
+    LoggingTransaction,
+    make_in_list_sql_clause,
+)
 from synapse.storage.databases.main.stream import generate_pagination_where_clause
 from synapse.storage.relations import (
     AggregationPaginationToken,
     PaginationChunk,
     RelationPaginationToken,
 )
 from synapse.util.caches.descriptors import cached
+from synapse.util.caches.lrucache import LruCache
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
 
 logger = logging.getLogger(__name__)
 
 
 class RelationsWorkerStore(SQLBaseStore):
+    def __init__(
+        self,
+        database: DatabasePool,
+        db_conn: LoggingDatabaseConnection,
+        hs: "HomeServer",
+    ):
+        super().__init__(database, db_conn, hs)
+
+        self._msc1849_enabled = hs.config.experimental.msc1849_enabled
+        self._msc3440_enabled = hs.config.experimental.msc3440_enabled
+
+        self.get_applicable_edit: LruCache[str, Optional[EventBase]] = LruCache(
+            cache_name="get_applicable_edit",
+            max_size=hs.config.caches.event_cache_size,  # TODO
+        )
```

Comment:
I'm not sure how we usually decide what to use the `max_size` in our caches? I guess this could be set to whatever the default for the `@cached` decorator is since we're replacing that.

Reply:
I'm also not 100% sure that this is truly interchangeable -- does the cache invalidation code work still with this?

Reply:
Unsure about these :(